src/frontend/src/utils/openID.ts

```diff
@@ -0,0 +1,195 @@
+import { Principal } from "@dfinity/principal";
+import { isNullish, nonNullish } from "@dfinity/utils";
+
+export interface RequestConfig {
+  // OAuth client ID
+  clientId: string;
+  // OAuth authentication URL
+  authURL: string;
+  // Optional, FedCM config URL
+  configURL?: string;
+}
+
+export interface RequestOptions {
+  // Nonce created from a principal with `createAnonymousNonce`
+  nonce: string;
+  // Optional, account identifier (e.g. email) used as login hint
+  loginHint?: string;
+  // Optional, see: https://developers.google.com/privacy-sandbox/blog/fedcm-auto-reauthn#mediation-options
+  mediation?: CredentialMediationRequirement;
+}
+
+export const GOOGLE_REQUEST_CONFIG: RequestConfig = {
+  clientId: II_OPENID_GOOGLE_CLIENT_ID,
+  authURL: "https://accounts.google.com/o/oauth2/v2/auth",
+  configURL: "https://accounts.google.com/gsi/fedcm.json",
+};
+
+/**
+ * Request JWT with the FedCM API
+ * @param config of the OpenID provider
+ * @param options for the JWT request
+ */
+const requestWithCredentials = async (
+  config: Omit<RequestConfig, "authURL">,
+  options: RequestOptions
+): Promise<string> => {
+  const identityCredential = await navigator.credentials.get({
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    identity: {
+      providers: [
+        {
+          configURL: config.configURL,
+          clientId: config.clientId,
+          nonce: options.nonce,
+          loginHint: options.loginHint,
+        },
+      ],
+    },
+    mediation: options.mediation,
+  });
+
+  if (
+    identityCredential?.type !== "identity" ||
+    !("token" in identityCredential) ||
+    typeof identityCredential.token !== "string"
+  ) {
+    // This should be unreachable in FedCM spec compliant browsers
+    throw new Error("Invalid credential received from FedCM API");
+  }
+
+  return identityCredential.token;
+};
+
+/**
+ * @param error to check whether it is a FedCM not supported error
+ */
+export const isNotSupportedError = (error: unknown) =>
+  error instanceof Error && error.name === "NotSupportedError";
+
+/**
+ * @param error to check whether it is a FedCM no permission error
+ */
+export const isPermissionError = (error: unknown) =>
+  error instanceof Error && error.name === "NetworkError";
+
+const toBase64 = (bytes: ArrayBuffer): string =>
+  btoa(String.fromCharCode(...new Uint8Array(bytes)));
+
+const toBase64URL = (bytes: ArrayBuffer): string =>
+  toBase64(bytes).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
+
+/**
+ * Request JWT through redirect flow in a popup
+ * @param config of the OpenID provider
+ * @param options for the JWT request
+ */
+const requestWithRedirect = async (
+  config: Omit<RequestConfig, "configURL">,
+  options: RequestOptions
+): Promise<string> => {
+  const state = toBase64URL(window.crypto.getRandomValues(new Uint8Array(12)));
+
+  const redirectURL = new URL(window.location.origin);
+  redirectURL.pathname = "/callback";
```

Comment:
Do we want to create a constant `/callback` to be used here and in the index.ts?